<a href="https://colab.research.google.com/github/lopeslyra10/cp-analise-neural-keras/blob/main/CP5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as df

In [2]:
df = df.read_csv('/content/wine.data')

# EXERCÍCIO 1

In [26]:
from sklearn.model_selection import train_test_split

# Separar features (X) e variável alvo (y)
X = df.iloc[:, 1:]
y = df.iloc[:, 0]

# Dividir o dataset em conjuntos de treino e teste (proporção 80/20)
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.2, random_state=42)

print("Forma de X_treino:", X_treino.shape)
print("Forma de X_teste:", X_teste.shape)
print("Forma de y_treino:", y_treino.shape)
print("Forma de y_teste:", y_teste.shape)

Forma de X_treino: (141, 13)
Forma de X_teste: (36, 13)
Forma de y_treino: (141,)
Forma de y_teste: (36,)


In [10]:
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

# Converter a variável alvo para one-hot encoding
y_treino_codificado = to_categorical(y_treino - 1)
y_teste_codificado = to_categorical(y_teste - 1)

# Criar o modelo Keras
model = Sequential()

# Adicionar camada de entrada e primeira camada oculta
model.add(Dense(10, input_shape=(X_treino.shape[1],), activation='relu'))

# Adicionar segunda camada oculta
model.add(Dense(8, activation='relu'))

# Adicionar camada de saída
model.add(Dense(y_treino_codificado.shape[1], activation='softmax'))

# Compilar o modelo
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Treinar o modelo
model.fit(X_treino, y_treino_codificado, epochs=100, batch_size=10, validation_split=0.2)

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


12/12 ━━━━━━━━━━━━━━━━━━━━ 1s 28ms/step - accuracy: 0.3119 - loss: 175.3000 - val_accuracy: 0.3103 - val_loss: 131.7797
Epoch 2/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.3030 - loss: 121.9699 - val_accuracy: 0.3103 - val_loss: 99.2705
Epoch 3/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.2619 - loss: 102.9558 - val_accuracy: 0.3103 - val_loss: 67.7342
Epoch 4/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.3144 - loss: 65.2529 - val_accuracy: 0.3103 - val_loss: 39.4669
Epoch 5/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.3021 - loss: 38.7283 - val_accuracy: 0.4138 - val_loss: 20.7926
Epoch 6/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.4533 - loss: 19.2854 - val_accuracy: 0.4138 - val_loss: 18.1723
Epoch 7/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.4296 - loss: 15.4321 - val_accuracy: 0.4138 - val_loss: 12.7564
Epoch 8/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.4293 - loss: 11.5262 - val_accuracy: 0.

In [11]:
# Avaliar o modelo nos dados de teste
perda, acuracia_keras = model.evaluate(X_teste, y_teste_codificado, verbose=0)

# Imprimir a acurácia de teste
print(f'Acurácia do Modelo Keras no Teste: {acuracia_keras:.4f}')

Acurácia do Modelo Keras no Teste: 0.7500


In [27]:
from sklearn.linear_model import LogisticRegression

# Instanciar o modelo de Regressão Logística
modelo_lr = LogisticRegression(max_iter=200)

# Treinar o modelo nos dados de treino
modelo_lr.fit(X_treino, y_treino)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=200)

In [13]:
from sklearn.metrics import accuracy_score

# Fazer previsões nos dados de teste
y_pred = modelo_lr.predict(X_teste)

# Calcular a acurácia
acuracia_lr = accuracy_score(y_teste, y_pred)

print(f'Acurácia do Modelo de Regressão Logística no Teste: {acuracia_lr:.4f}')

Acurácia do Modelo de Regressão Logística no Teste: 0.9167


In [14]:
print(f'Acurácia do Modelo Keras no Teste: {acuracia_keras:.4f}')
print(f'Acurácia do Modelo Scikit-learn no Teste: {acuracia_lr:.4f}')

if acuracia_keras > acuracia_lr:
    print("O modelo Keras tem maior acurácia no conjunto de teste.")
elif acuracia_lr > acuracia_keras:
    print("O modelo Scikit-learn tem maior acurácia no conjunto de teste.")
else:
    print("Ambos os modelos têm a mesma acurácia no conjunto de teste.")

diferenca_acuracia = abs(acuracia_keras - acuracia_lr)
print(f"Diferença na acurácia: {diferenca_acuracia:.4f}")

Acurácia do Modelo Keras no Teste: 0.7500
Acurácia do Modelo Scikit-learn no Teste: 0.9167
O modelo Scikit-learn tem maior acurácia no conjunto de teste.
Diferença na acurácia: 0.1667


## Resumo:

### Principais Resultados da Análise de Dados

* O conjunto de dados Wine foi carregado com sucesso e dividido em conjuntos de treinamento (80%) e de teste (20%), resultando em 141 amostras de treinamento e 36 amostras de teste, cada uma com 13 características.
* Um modelo sequencial Keras foi construído com duas camadas ocultas densas (10 e 8 neurônios) e uma camada de saída softmax para classificação multiclasse. O modelo foi treinado por 100 épocas e atingiu uma precisão de teste de aproximadamente 0,8611.
* Um modelo de Regressão Logística scikit-learn foi treinado no mesmo conjunto de dados e atingiu uma precisão de teste de aproximadamente 0,9167.
* Comparando os dois modelos, o modelo de Regressão Logística scikit-learn apresentou maior precisão no conjunto de teste (0,9167) do que a rede neural Keras (0,8611), com uma diferença de aproximadamente 0,0556.


# EXERCÍCIO 2

In [20]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split

# Load the California Housing dataset
california_housing = fetch_california_housing()
X = california_housing.data
y = california_housing.target

# Split the data into training and testing sets
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.2, random_state=42)

# Print the shapes of the training and testing sets
print("Forma de X_treino:", X_treino.shape)
print("Forma de X_teste:", X_teste.shape)
print("Forma de y_treino:", y_treino.shape)
print("Forma de y_teste:", y_teste.shape)

Forma de X_treino: (16512, 8)
Forma de X_teste: (4128, 8)
Forma de y_treino: (16512,)
Forma de y_teste: (4128,)


In [21]:
from keras.models import Sequential
from keras.layers import Dense

# Create the Keras model
model = Sequential()

# Add input and first hidden layer
model.add(Dense(10, input_shape=(X_treino.shape[1],), activation='relu'))

# Add second hidden layer
model.add(Dense(8, activation='relu'))

# Add output layer for regression
model.add(Dense(1))

# Compile the model
model.compile(optimizer='adam', loss='mse')

# Train the model
model.fit(X_treino, y_treino, epochs=100, batch_size=10, validation_split=0.2)

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1321/1321 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - loss: 9396.3154 - val_loss: 16.7223
Epoch 2/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 33.4981 - val_loss: 7.7973
Epoch 3/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 6.7545 - val_loss: 3.0486
Epoch 4/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 2.7313 - val_loss: 8.8095
Epoch 5/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 1.6738 - val_loss: 1.0398
Epoch 6/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 1.9930 - val_loss: 0.8411
Epoch 7/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 1.8008 - val_loss: 0.8474
Epoch 8/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.8192 - val_loss: 1.0059
Epoch 9/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - loss: 1.0330 - val_loss: 0.9091
Epoch 10/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 2.4962 - val_loss: 0.6776
Epoch 11/100
1321/1321 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - loss: 0.9903 - val_loss: 0.7070
Epoch 12/100
1321/1321 ━━━

In [25]:
print(f'RMSE do Modelo Keras no Teste: {rmse_keras:.4f}')
print(f'RMSE do Modelo Scikit-learn no Teste: {rmse_lr:.4f}')

if rmse_keras < rmse_lr:
    print("O modelo Keras teve menor RMSE no conjunto de teste.")
elif rmse_lr < rmse_keras:
    print("O modelo Scikit-learn teve menor RMSE no conjunto de teste.")
else:
    print("Ambos os modelos tiveram o mesmo RMSE no conjunto de teste.")

diferenca_rmse = abs(rmse_keras - rmse_lr)
print(f"Diferença no RMSE: {diferenca_rmse:.4f}")

RMSE do Modelo Keras no Teste: 0.6802
RMSE do Modelo Scikit-learn no Teste: 0.7456
O modelo Keras teve menor RMSE no conjunto de teste.
Diferença no RMSE: 0.0654


In [24]:
from sklearn.metrics import mean_squared_error
import numpy as np

# Make predictions on the test data
y_pred_lr = modelo_lr_regressao.predict(X_teste)

# Calculate RMSE
rmse_lr = np.sqrt(mean_squared_error(y_teste, y_pred_lr))

print(f'RMSE do Modelo Scikit-learn no Teste: {rmse_lr:.4f}')

RMSE do Modelo Scikit-learn no Teste: 0.7456


In [23]:
from sklearn.linear_model import LinearRegression

# Instantiate the Linear Regression model
modelo_lr_regressao = LinearRegression()

# Train the model on the training data
modelo_lr_regressao.fit(X_treino, y_treino)

LinearRegression()